diff --git a/blis b/blis
--- a/blis
+++ b/blis
@@ -1 +1 @@
-Subproject commit a0b04e3c007f1207e5678bf20c07752906742fb7
+Subproject commit a0b04e3c007f1207e5678bf20c07752906742fb7-dirty
diff --git a/lib/ExecutionEngine/GpuUtils.cpp b/lib/ExecutionEngine/GpuUtils.cpp
index a73f98b..32e802b 100644
--- a/lib/ExecutionEngine/GpuUtils.cpp
+++ b/lib/ExecutionEngine/GpuUtils.cpp
@@ -1,5 +1,6 @@
 #include <cstddef>
 #include <cstdint>
+#include <ctime>
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
@@ -154,6 +155,11 @@ extern "C" __attribute__((visibility("default"))) void cudaLaunchKernel(int64_t
     // }
 
     free(name_with_suffix);
+    cuCtxSynchronize();
 
-    CU_CHECK(cuLaunchKernel(cuFunction, blocksPerGridX, blocksPerGridY, blocksPerGridZ, numWraps* threadsPerWarp, 1, 1, sharedMem, 0, cast_args, 0));
+    for(int i = 0; i < 10; i++)
+    {
+        CU_CHECK(cuLaunchKernel(cuFunction, blocksPerGridX, blocksPerGridY, blocksPerGridZ, numWraps* threadsPerWarp, 1, 1, sharedMem, 0, cast_args, 0));
+    }
+    cuCtxSynchronize();
 }
\ No newline at end of file
diff --git a/triton b/triton
--- a/triton
+++ b/triton
@@ -1 +1 @@
-Subproject commit 6fc969d896bd65e48b2dba1394d92b542652e5a4
+Subproject commit 6fc969d896bd65e48b2dba1394d92b542652e5a4-dirty
