diff --git a/CMakeLists.txt b/CMakeLists.txt
index 309855e0c..b2530e31c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -103,7 +103,7 @@ endfunction()
 
 
 # Disable warnings that show up in external code (gtest;pybind11)
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror -Wno-covered-switch-default -fvisibility=hidden")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-covered-switch-default -fvisibility=hidden")
 
 include_directories(".")
 include_directories(${MLIR_INCLUDE_DIRS})
@@ -220,6 +220,6 @@ if(TRITON_BUILD_PYTHON_MODULE AND NOT WIN32)
   target_link_libraries(triton PRIVATE ${PYTHON_LDFLAGS})
 endif()
 
-add_subdirectory(bin)
-add_subdirectory(test)
-add_subdirectory(unittest)
+# add_subdirectory(bin)
+# add_subdirectory(test)
+# add_subdirectory(unittest)
diff --git a/include/triton/Dialect/NVGPU/IR/CMakeLists.txt b/include/triton/Dialect/NVGPU/IR/CMakeLists.txt
index f8932cdc4..37fb66417 100644
--- a/include/triton/Dialect/NVGPU/IR/CMakeLists.txt
+++ b/include/triton/Dialect/NVGPU/IR/CMakeLists.txt
@@ -1,8 +1,8 @@
 set(MLIR_BINARY_DIR ${CMAKE_BINARY_DIR})
 
 set(LLVM_TARGET_DEFINITIONS NVGPUOps.td)
-mlir_tablegen(Dialect.h.inc -gen-dialect-decls -dialect=nvgpu)
-mlir_tablegen(Dialect.cpp.inc -gen-dialect-defs -dialect=nvgpu)
+mlir_tablegen(Dialect.h.inc -gen-dialect-decls -dialect=ttnvgpu)
+mlir_tablegen(Dialect.cpp.inc -gen-dialect-defs -dialect=ttnvgpu)
 mlir_tablegen(OpsConversions.inc -gen-llvmir-conversions)
 mlir_tablegen(Ops.h.inc -gen-op-decls)
 mlir_tablegen(Ops.cpp.inc -gen-op-defs)
diff --git a/include/triton/Dialect/NVGPU/IR/NVGPUDialect.td b/include/triton/Dialect/NVGPU/IR/NVGPUDialect.td
index 6978173d4..3b76e10c7 100644
--- a/include/triton/Dialect/NVGPU/IR/NVGPUDialect.td
+++ b/include/triton/Dialect/NVGPU/IR/NVGPUDialect.td
@@ -25,7 +25,7 @@
 include "mlir/IR/OpBase.td"
 
 def NVGPU_Dialect : Dialect {
-  let name = "nvgpu";
+  let name = "ttnvgpu";
   let cppNamespace = "::mlir::triton::nvgpu";
 
   let description = [{
